### ISSUE TITLE
Preprocessing more than 4 channels image

### ISSUE THREAD

**Issue by PPACI** [fchollet/keras#4731]
Please make sure that the boxes below are checked before you submit your issue. Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

---------------------------
Hi Keras.
I have a pipeline where i need to preprocess an image with 5 channels (random rotation/shearing). So the dimensions of an image is (x, y, 5) in 'tf' ordering mode. It was working nicely until i updated to latest keras version.
Then i got an error saying that the image must must be either 1, 3 or 4 channels on last axis.

I found a quickfix by commenting those line in keras.preprocessing.image : 
`if self.X.shape[channels_axis] not in {1, 3, 4}:
            raise ValueError('NumpyArrayIterator is set to use the '
                             'dimension ordering convention "' + dim_ordering + '" '
                             '(channels on axis ' + str(channels_axis) + '), i.e. expected '
                             'either 1, 3 or 4 channels on axis ' + str(channels_axis) + '. '
                             'However, it was passed an array with shape ' + str(self.X.shape) +
                             ' (' + str(self.X.shape[channels_axis]) + ' channels).')`
Around line 550 in the files.
So my question is, is there a way to deal with more than 4 channel images ? Have i broken something in my local keras ?

Thank you all

**Comment by fchollet** [comment]
Curious, how comes you have images with 5 channels?

**Comment by PPACI** [comment]
I work on an semantic segmentation model with the output being a image with one channel for each class. 
My input data is like (x, 128,128, 3) : a classic set of 128.128 RGB image
My output data is like (x, 128, 128, 5) : a fake image semantically segmented on each layer

You aroused my curiosity, based on paper i read, the N channels output seemed to be the good way to go but is there another way ?

**Comment by fchollet** [comment]
In this case, I don't understand why you would need `keras.preprocessing.image` to preprocess your outputs? This module is meant specifically for preprocessing of natural images (augmentation, normalization).

**Comment by PPACI** [comment]
The target in my training set is a mask (with 5 channel).
I preprocess input image with random rotation and shearing, then preprocess the mask. I should mention that I instantiate one generator and then call twice .flow(), passing the input then the mask with the same seed.
I don't know if I'm understandable, if not I will insert code example later.

I think that using this kind of mask as target is not a best practice since dealing with X channel image is unnatural but it looked like the easiest way to implement a segment of segmentation model. And the output of the network is very satisfying indeed.

**Comment by grantbey** [comment]
I'd like to bring this back to life - I'm doing semantic image segmentation of satellite imagery, and there are 20 channels. I'd really like for the generator to be able to handle this. As of now, I'm doing all my segmentation beforehand.

**Comment by juggernaut5k** [comment]
Commenting out lines 771 to 777 was a quick fix for me. I didn't notice any ill effects by doing so after writing out the augmented files to disk to check. Additionally had to add this to the next method part of the NumpyArrayIterator class.

        if self.save_to_dir:
            for i in range(current_batch_size):
                fname = '{prefix}_{index}_{hash}.{format}'.format(prefix=self.save_prefix,
                                                                                      index=current_index + i,
                                                                                      hash=np.random.randint(1e4),
                                                                                      format=self.save_format)
                if self.save_format == 'npy':
                    np.save(os.path.join(self.save_to_dir, fname), batch_x[i])
                else:
                    img = array_to_img(batch_x[i], self.data_format, scale=True)
                    img.save(os.path.join(self.save_to_dir, fname))

This was simply my hot fix for using 6 satellite bands with augmentation quickly. The 4 channel limit seems to be a limitation of the output file types rather than anything to do with the augmentation itself.

**Comment by lorenzori** [comment]
on v 2.1.6 the trick above does not work (code has changed). I would need 6 channels, anyone sorted it out already?

**Comment by RunningDongxu** [comment]
Interesting, does anyone have a solution?
